File: 1351-count-negative-numbers-in-a-sorted-matrix/1351-count-negative-numbers-in-a-sorted-matrix.py
class Solution:
    def countNegatives(self, grid: List[List[int]]) -> int:
        res = 0

        for g in grid:
            n = len(g)

            if g[0] < 0:
                res += n
                continue

            l, r = 0, n - 1
            while l <= r:
                mid = (l + r) // 2
                if g[mid] < 0:
                    r = mid - 1
                else:          # g[mid] >= 0
                    l = mid + 1

            # l is now the index of the first negative
            res += n - l

        return res

File: 1351-count-negative-numbers-in-a-sorted-matrix/1351-count-negative-numbers-in-a-sorted-matrix.cpp
class Solution {
public:
    int countNegatives(vector<vector<int>>& grid) {
        int res = 0;

        for (auto& row : grid) {
            int l = 0, r = row.size() - 1;
            int ans = row.size();   // index of first negative

            while (l <= r) {
                int mid = l + (r - l) / 2;

                if (row[mid] < 0) {
                    ans = mid;
                    r = mid - 1;
                } else {
                    l = mid + 1;
                }
            }
            res += row.size() - ans;
        }

        return res;
    }
};